File: Fs.Units.CSharp/Simple/Frequency.cs
// ReSharper disable once CheckNamespace
namespace Fs.Units.CSharp.Frequency;

public static class HertzExtensions
{
    public static double HertzToKilohertz(this double x) => Hertz.toKilohertz(x);
    public static double HertzToMegahertz(this double x) => Hertz.toMegahertz(x);
    public static double HertzToGigahertz(this double x) => Hertz.toGigahertz(x);
}

public static class KilohertzExtensions
{
    public static double KilohertzToHertz(this double x) => Kilohertz.toHertz(x);
    public static double KilohertzToMegahertz(this double x) => Kilohertz.toMegahertz(x);
    public static double KilohertzToGigahertz(this double x) => Kilohertz.toGigahertz(x);
}

public static class MegahertzExtensions
{
    public static double MegahertzToHertz(this double x) => Megahertz.toHertz(x);
    public static double MegahertzToKilohertz(this double x) => Megahertz.toKilohertz(x);
    public static double MegahertzToGigahertz(this double x) => Megahertz.toGigahertz(x);
}

public static class GigahertzExtensions
{
    public static double GigahertzToHertz(this double x) => Gigahertz.toHertz(x);
    public static double GigahertzToKilohertz(this double x) => Gigahertz.toKilohertz(x);
    public static double GigahertzToMegahertz(this double x) => Gigahertz.toMegahertz(x);
}

File: src/Fs.Units/Simple/Frequency.fs
namespace Fs.Units

[<AutoOpen>]
module Frequency =

    /// Hertz
    [<Measure>]
    type Hz

    /// Kilohertz
    [<Measure>]
    type kHz

    /// Megahertz
    [<Measure>]
    type MHz

    /// Gigahertz
    [<Measure>]
    type GHz

    [<RequireQualifiedAccess>]
    module Hertz =

        let create (x: float) = x * 1.0<Hz>
        let toKilohertz (x: float<Hz>) = x / 1e+3<Hz> * 1.0<kHz>
        let toMegahertz (x: float<Hz>) = x / 1e+6<Hz> * 1.0<MHz>
        let toGigahertz (x: float<Hz>) = x / 1e+9<Hz> * 1.0<GHz>

    [<RequireQualifiedAccess>]
    module Kilohertz =

        let create (x: float) = x * 1.0<kHz>
        let toHertz (x: float<kHz>) = x * 1e+3<Hz> / 1.0<kHz>
        let toMegahertz (x: float<kHz>) = x / 1e+3<kHz> * 1.0<MHz>
        let toGigahertz (x: float<kHz>) = x / 1e+6<kHz> * 1.0<GHz>

    [<RequireQualifiedAccess>]
    module Megahertz =

        let create (x: float) = x * 1.0<MHz>
        let toHertz (x: float<MHz>) = x * 1e+6<Hz> / 1.0<MHz>
        let toKilohertz (x: float<MHz>) = x * 1e+3<kHz> / 1.0<MHz>
        let toGigahertz (x: float<MHz>) = x / 1e+3<MHz> * 1.0<GHz>

    [<RequireQualifiedAccess>]
    module Gigahertz =

        let create (x: float) = x * 1.0<GHz>
        let toHertz (x: float<GHz>) = x * 1e+9<Hz> / 1.0<GHz>
        let toKilohertz (x: float<GHz>) = x * 1e+6<kHz> / 1.0<GHz>
        let toMegahertz (x: float<GHz>) = x * 1e+3<MHz> / 1.0<GHz>
